app/bridge/smuggler.js

```diff
@@ -154,7 +159,32 @@
     registerContract(name);
 
     var registration = registrations.get(name);
-    registration.clients.push(uuid);
+    if (!clientAlreadyRegistered(uuid, name)) {
+      registration.clients.push(uuid);
+    }
+  }
+
+  function clientAlreadyRegistered(clientId, name) {
+    for (var client of getClientsForContract(name)) {
+      if (clientId === client) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  function unregisterClientForContract(uuid, name) {
+    var registration = registrations.get(name);
+    if (registration) {
+      var index = registration.clients.indexOf(uuid);
+      if (index > -1) {
+        registration.clients.splice(index, 1);
+      } else {
+        debug('Cannot remove non-existing client ' + uuid + ' from ' + name);
+      }
+    } else {
+      debug('Cannot remove client from non existing contract ' + name);
+    }
```

Comment:
It usually make less indentation and easier code to read (because of less nesting) to do:

if (!registration) {
  debug('.....');
  return;
}

var index = registration.clients.indexOf(uuid);
if (index === -1) {
  debug('...');
  return;
}

registration.clients.splice(index, 1);